src/main.rs

```diff
@@ -30,82 +31,126 @@ mod error;
 mod path;
 
 use midi::PortMidiNoteTracker;
+use pm::PortMidiDeviceId as DeviceId;
 use ui::Popup;
 use screen::*;
 use config::Config;
 use hardcode::*;
 use error::Result;
 
-fn print_devices(pm: &pm::PortMidi) {
-    for dev in pm.devices().unwrap() {
-        println!("{}", dev);
-    }
-}
-
 fn config_path() -> Result<PathBuf> {
     env::home_dir()
         .ok_or("Home directory not found".into())
         .map(|config_dir| config_dir.join(hardcode::CONFIG_FILE_NAME))
 }
 
-fn main() {
-    let context = pm::PortMidi::new().unwrap();
+type Looper = looper::Looper<PortMidiNoteTracker>;
 
-    let (input_id, output_id) = {
-        let args: Vec<String> = std::env::args().collect();
-
-        if args.len() < 2 {
-            print_devices(&context);
-            println!("Usage: ./dimooper <input-port> <output-port>");
-            std::process::exit(1);
-        }
-
-        (args[1].trim().parse().unwrap(), args[2].trim().parse().unwrap())
-    };
-
-    let in_info = context.device(input_id).unwrap();
-    println!("Listening on: {} {}", in_info.id(), in_info.name());
-    let in_port = context.input_port(in_info, 1024).unwrap();
-
-    let out_info = context.device(output_id).unwrap();
+fn create_looper(context: &pm::PortMidi, output_id: DeviceId) -> Result<Looper> { 
+    let out_info = try!(context.device(output_id));
     println!("Sending recorded events: {} {}",
              out_info.id(),
              out_info.name());
-    let out_port = context.output_port(out_info, 1024).unwrap();
+    let out_port = try!(context.output_port(out_info, 1024));
+    let looper = looper::Looper::new(PortMidiNoteTracker::new(out_port));
+    Ok(looper)
+}
 
-    let window_width = RATIO_WIDTH * RATIO_FACTOR;
+fn create_event_loop(context: &pm::PortMidi, input_id: DeviceId) -> Result<EventLoop<'static>> {
+    let window_width = RATIO_WIDTH  * RATIO_FACTOR;
     let window_height = RATIO_HEIGHT * RATIO_FACTOR;
-    let sdl_context = sdl2::init().unwrap();
-    let video_subsystem = sdl_context.video().unwrap();
-    let timer_subsystem = sdl_context.timer().unwrap();
-    let ttf_context = sdl2_ttf::init().unwrap();
 
-    let window = video_subsystem.window("Dimooper", window_width, window_height)
+    let in_info = try!(context.device(input_id));
+    println!("Listening on: {} {}", in_info.id(), in_info.name());
+    let in_port = try!(context.input_port(in_info, 1024));
+
+    let sdl_context = try!(sdl2::init());
+    let video_subsystem = try!(sdl_context.video());
+    let timer_subsystem = try!(sdl_context.timer());
```

Comment:
I don't like that the creation of `EventLoop` is coupled with SDL initialization. `EventLoop` was design as a first class entity and I will probably create several instances of it in different places. So this code should be decoupled.

**This remark doesn't block the merge. Feel free to not fix it. I'm gonna file it as a techdebt to fix later.**